tools/flang1/flang1exe/dtypeutl.c

```diff
@@ -2115,6 +2115,50 @@ same_dtype(DTYPE d1, DTYPE d2)
   return FALSE;
 }
 
+/** \brief Similar to eq_dtype(), except all scalar integer types are compatible
+    with each other, all scalar logical types are compatible with each other,
+    all character types are compatible with each other and we allow
+    casting arrays when types fit into each other.
+ */
+LOGICAL
+cmpat_dtype_array_cast(DTYPE d1, DTYPE d2)
+{
+  int s1, s2;
+
+  if (d1 == d2)
+    return TRUE;
+  if (DTY(d1) != DTY(d2)) {
+    /* check for any logical first since logical types also have the
+     * _TY_INT attribute.
+     */
+    if (DT_ISLOG(d1) && DT_ISLOG(d2))
+      return TRUE;
+    if (DT_ISINT(d1) && DT_ISINT(d2))
+      return TRUE;
+    return FALSE;
+  }
+  if (DTY(d1) == TY_CHAR || DTY(d1) == TY_NCHAR)
+    return TRUE;
+
+  /* Perform checks for array casting possibilities */
+  if (DTY(d1) == TY_ARRAY && DTY(d2) == TY_ARRAY) {
```

Comment:
Since we know `DTY(d1) == DTY(d2)`, why check both of them here?

Reply:
True, I removed the second check and added a comment explaining that we are already in a place where we know that the types are identical.